File: Node/JsonPacketSerialization.cpp
#include "JsonPacketSerialization.hpp"

/*
MESSAGE FORMAT: CODE(1 byte) MESSAGE LENGTH(4 bytes) JSON MESSAGE(n bytes)
*/

/*
* this function creates a json reprersentation in text that contains the login response message
* input: login response
* output: json in text
*/
unsigned char* JsonPacketSerialization::createLoginResult(LoginResult loginResult){
	//create json message
	json j;
	j["code"] = loginResult.code;

	return getJsonBufferResponse(j);
}

/*
* this function creates a json reprersentation in text that contains the signup response message
* input: signup response
* output: json in text
*/
unsigned char* JsonPacketSerialization::createSignupResult(SignupResult signupResult){
	json j;
	j["code"] = signupResult.code;

	return getJsonBufferResponse(j);
}

/*
* this function creates a json reprersentation in text that contains the error result message
* input: error result
* output: json in text
*/
unsigned char* JsonPacketSerialization::createErrorRequestResult(ErrorResult errorResult){
	json j;
	j["message"] = errorResult.errorMessage;
	j["code"] = errorResult.code;

	return getJsonBufferResponse(j);
}

/*
* this function creates a json reprersentation in text that contains the amount of connected users result message
* input: amount of connected users result
* output: json in text
*/
unsigned char* JsonPacketSerialization::createUpdateUsersResult(int amountOfConnectedUsers){
	json j;
	j["connectedUsers"] = amountOfConnectedUsers;

	return getJsonBufferResponse(j);
}

/*
* this function creates a text representation of a json
* input: json
* output: text representation
*/
unsigned char* JsonPacketSerialization::getJsonBufferResponse(json j)
{
	unsigned char* buffer;
	//getting json string
	std::string jsonString = j.dump();
	//create a buffer to the response
	buffer = new unsigned char[BYTES_LEN + jsonString.length() + 1];

	// getting the json string length in 4 byte format and add it to the buffer
	unsigned char* byteLen = Helper::changeIntTo4ByteCharArray(jsonString.length());
	//add the bytes to the buffer
	for (int i = 1; i < BYTES_LEN + 1; i++) {
		buffer[i] = byteLen[i-1];
	}
	delete byteLen; //clean memory

	//add the json string to the buffer
	for (int i = 1; i < jsonString.length() + 1; i++) {
		buffer[i + BYTES_LEN] = (unsigned char)jsonString[i-1];
	}
	return buffer;
}

File: PythonServer/JsonPacketSerialization.py
import json

class JsonPacketSerialization:
    def __init__(self):
        pass
    
    """
    this function creates a json buffer fron user input
    input: code of the login messsage and the user id
    output: json buffer
    """
    @staticmethod
    def get_login_result(code, user_id):
        response_json = {
            "code": code,
            "user_id": user_id
        }
        return JsonPacketSerialization.__createResponseBuffer(response_json)

    """
    this function creates a json buffer fron user input
    input: code of the signup messsage 
    output: json buffer
    """
    @staticmethod 
    def get_signup_result(code):
        response_json = {
            "code": code
        }
        return JsonPacketSerialization.__createResponseBuffer(response_json)
    
    """
    Method that create a buffer by a json object.
    The method create a buffer with the message length and the json string in bytes format.
    Input: json object.
    Output: bytes buffer.
    """
    @staticmethod
    def __createResponseBuffer(response_json):
        # convert the json object to string
        json_string = json.dumps(response_json).replace("'", '"')

        # getting the length of the json string
        json_length = len(json_string)
        
        # getting the json string length in byte format
        json_length_bytes_string = JsonPacketSerialization.__int_to_4_bytes_string(json_length)

        # create a buffer in byte format that have the message length and the json string
        buffer = json_length_bytes_string + json_string
        return buffer


    """
    Method that convert int to 4 bytes array.
    Input: number to convert.
    Output: 4 byte array.
    """
    @staticmethod
    def __int_to_4_bytes_string(num):
        result_bytes = []
        result = ""

        # create array of bytes
        for i in range(0, 4):
            result_bytes.append(num >> (i * 8) & 0xff)
        result_bytes.reverse()
        
        # convert the bytes array to char array
        for i in range(0, 4):
            result += chr(result_bytes[i])
        
        return result
